File: app/src/main/java/com/example/movies/ui/trending/view/TabPersonFragment.java
package com.example.movies.ui.trending.view;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import com.example.movies.R;
import com.example.movies.core.base.BaseFragment;
import com.example.movies.data.entity.Person;
import com.example.movies.data.entity.Result;
import com.example.movies.data.entity.TabPersonEntity;
import com.example.movies.data.service.RetroFitService;
import com.example.movies.ui.trending.adapter.TabPersonAdapter;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.Unbinder;


public class TabPersonFragment extends BaseFragment {
    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;

    private TabPersonAdapter adapter;
    private ArrayList<Person> people = new ArrayList<>();
    private GridLayoutManager layoutManager;
    private boolean isLoadingShowed = false;
    private int pagination = 1;
    private int sizeOfPage = 0;
    RetroFitService service = new RetroFitService();

    @Override
    public Integer getFragmentLayoutId() {
        return R.layout.fragment_layout;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setRecyclerView();
        initScrollListener();
        fetchPerson();
    }

    private void initScrollListener() {
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (!isLoadingShowed) {
                    if (layoutManager != null && layoutManager.findLastCompletelyVisibleItemPosition() == people.size() - 1) {
                        loadMore();
                        isLoadingShowed = true;
                    }
                }
            }
        });
    }

    private void loadMore() {
        pagination++;
        people.add(null);
        adapter.notifyDataSetChanged();
        fetchPerson();
    }

    private void fetchPerson() {
        RetroFitService.ResultCallBack serviceCallBack = new RetroFitService.ResultCallBack() {
            @Override
            public void getResult(Result result) {
                Result<TabPersonEntity> tabPersonEntityResult = result;
                if (tabPersonEntityResult.getData() != null ||
                        (tabPersonEntityResult.getData() != null ? tabPersonEntityResult.getData().getResults() : null) != null
                                && (pagination == 1 || sizeOfPage == tabPersonEntityResult.getData().getResults().size())) {
                    setData(tabPersonEntityResult.getData().getResults());
                } else {
                    setLoadingCase();
                }
            }
        };
        service.fetchPerson(serviceCallBack);
    }

    private void setRecyclerView() {
        layoutManager = new GridLayoutManager(getContext(), 2);
        recyclerView.setLayoutManager(layoutManager);
        adapter = new TabPersonAdapter(getContext(), people);
        recyclerView.setAdapter(adapter);
    }

    private void setLoadingCase() {
        if (isLoadingShowed) {
            people.remove(people.size() - 1);
            adapter.notifyDataSetChanged();
            isLoadingShowed = false;
        }
    }

    private void addItemsToList(ArrayList<Person> list) {
        sizeOfPage = list.size();
        people.addAll(list);
        adapter.notifyDataSetChanged();
    }

    private void setData(ArrayList<Person> list) {
        setLoadingCase();
        addItemsToList(list);
    }

}

File: app/src/main/java/com/example/movies/ui/trending/view/TabPersonFragment.kt
package com.example.movies.ui.trending.view

import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import com.example.movies.R
import com.example.movies.core.base.BaseFragment
import com.example.movies.data.entity.Person
import com.example.movies.data.entity.Result
import com.example.movies.data.entity.TabPersonEntity
import com.example.movies.data.service.RetroFitService
import com.example.movies.data.service.RetroFitService.ResultCallBack
import com.example.movies.ui.trending.adapter.TabPersonAdapter
import kotlinx.android.synthetic.main.fragment_layout.*
import java.util.*

class TabPersonFragment : BaseFragment() {
    private var adapter: TabPersonAdapter? = null
    private val people = ArrayList<Person?>()
    private var layoutManager: GridLayoutManager? = null
    private var isLoadingShowed = false
    private var pagination = 1
    private var sizeOfPage = 0
    var service = RetroFitService()
    override fun getFragmentLayoutId(): Int {
        return R.layout.fragment_layout
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setRecyclerView()
        initScrollListener()
        fetchPerson()
    }

    private fun initScrollListener() {
        recyclerView!!.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
            }

            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (!isLoadingShowed) {
                    if (layoutManager != null && layoutManager!!.findLastCompletelyVisibleItemPosition() == people.size - 1) {
                        loadMore()
                        isLoadingShowed = true
                    }
                }
            }
        })
    }

    private fun loadMore() {
        pagination++
        people.add(null)
        adapter!!.notifyDataSetChanged()
        fetchPerson()
    }

    private fun fetchPerson() {
        val serviceCallBack: ResultCallBack = object : ResultCallBack {
            override fun getResult(result: Result<*>) {
                val tabPersonEntityResult: Result<TabPersonEntity> = result as Result<TabPersonEntity>
                if (tabPersonEntityResult.data != null ||
                        (if (result.data != null) tabPersonEntityResult.data!!.results else null) != null
                        && (pagination == 1 || sizeOfPage == tabPersonEntityResult.data!!.results!!.size)) {
                    setData(result.data!!.results)
                } else {
                    setLoadingCase()
                }
            }
        }
        service.fetchPerson(serviceCallBack)
    }

    private fun setRecyclerView() {
        layoutManager = GridLayoutManager(context, 2)
        recyclerView!!.layoutManager = layoutManager
        adapter = TabPersonAdapter(context!!, people)
        recyclerView!!.adapter = adapter
    }

    private fun setLoadingCase() {
        if (isLoadingShowed) {
            people.removeAt(people.size - 1)
            adapter!!.notifyDataSetChanged()
            isLoadingShowed = false
        }
    }

    private fun addItemsToList(list: ArrayList<Person>?) {
        sizeOfPage = list!!.size
        people.addAll(list)
        adapter!!.notifyDataSetChanged()
    }

    private fun setData(list: ArrayList<Person>?) {
        setLoadingCase()
        addItemsToList(list)
    }
}